server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -178,34 +183,86 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
                         }
                     }
                     if (indexMetaData != null) {
-                        // Find the default pipeline if one is defined from and existing index.
-                        String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexMetaData.getSettings());
-                        indexRequest.setPipeline(defaultPipeline);
-                        if (IngestService.NOOP_PIPELINE_NAME.equals(defaultPipeline) == false) {
-                            hasIndexRequestsWithPipelines = true;
+                        final Settings indexSettings = indexMetaData.getSettings();
+                        if (IndexSettings.REQUIRED_PIPELINE.exists(indexSettings)) {
+                            // find the required pipeline if one is defined from an existing index
+                            requiredPipeline = IndexSettings.REQUIRED_PIPELINE.get(indexSettings);
+                            assert IndexSettings.DEFAULT_PIPELINE.get(indexSettings).equals(IngestService.NOOP_PIPELINE_NAME) :
+                                IndexSettings.DEFAULT_PIPELINE.get(indexSettings);
+                            indexRequest.setPipeline(requiredPipeline);
+                            requestCanOverridePipeline = false;
+                        } else {
+                            // find the default pipeline if one is defined from an existing index
+                            String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexSettings);
+                            indexRequest.setPipeline(defaultPipeline);
```

Comment:
Should we check whether `requestPipeline` is not null here, otherwise a default pipeline would overwrite the specified pipeline on an index request and I think before this wouldn't happen? (before there was a `if (indexRequest != null) {` check in the beginning of the for loop body)

Reply:
It will get set back to the requestPipeline on line 243.  I did a quick test and the request pipeline still overrides the default pipeline (and provides a meaninful error for required pipeline)

Reply:
I misunderstood the code here. So if there is a default pipeline then it will be passed to `IndexRequest#setPipeline(...)`, but then later if the `requestedPipeline` is not null then it that variable is provided to `IndexRequest#setPipeline(...)`. So things work as expected.

In the current code, we skip checking for default pipeline entirely if there is a pipeline specified in the index request. I understand we can't do this now with required pipeline. We always need to check if it has been provided, whereas before we didn't always have to check IndexTemplateMetaData and IndexMetaData.